### PR TITLE
PVA examiner example

### LINKED ISSUE #60: show example PVA client
APS beam line scientist asks for a simple Python demo showing how to receive streaming PVA data:

> Is there any module or do you have any simple example of consumer receiver program for streaming data within BlueSky - could be with underlaying PVA (or PVAPY)?

### PR DESCRIPTION
**Pull request by prjemian**
- close #60

### CODE REVIEW

**Comment by prjemian** [comment]
Anyone care to take a look at this?  reviewers? @stevehenke?

**Comment by prjemian** [comment]
Here's some example output:

```py
(base) prjemian@zap:~/.../BCDA-APS/bluesky_training$ /home/prjemian/.conda/envs/bluesky_2023_2/bin/python /home/prjemian/Documents/projects/BCDA-APS/bluesky_training/bluesky/user/dev_qs_examiner.py
channel.getName()='ad:Pva1:Image'  channel.isConnected() = False
PVA monitor event: 'ad:Pva1:Image' at 2023-03-29 15:54:37.938574
image.shape=(1024, 1024)  image.dtype=dtype('uint8')  image.min()=9  image.max()=200
================ ====== ============================================================
key              length value                                                       
================ ====== ============================================================
alarm            43     {'severity': 0, 'status': 0, 'message': ''}                 
attribute        1335   [{'name': 'ColorMode', 'value': ({'value': 0}, {'value': pva
codec            87     {'name': '', 'parameters': ({'value': 5}, {'value': pvaccess
compressedSize   7      1048576                                                     
dataTimeStamp    72     {'secondsPastEpoch': 1680123277, 'nanoseconds': 938573718, '
descriptor       0                                                                  
dimension        158    [{'size': 1024, 'offset': 0, 'fullSize': 1024, 'binning': 1,
display          81     {'limitLow': 0.0, 'limitHigh': 0.0, 'description': '', 'form
timeStamp        72     {'secondsPastEpoch': 1680123277, 'nanoseconds': 961849218, '
uncompressedSize 7      1048576                                                     
uniqueId         6      731420                                                      
value            121    ({'ubyteValue': array([16, 21, 10, ..., 17, 20,  9], dtype=u
================ ====== ============================================================

NDAttributes (metadata)
=================== ==================================================================================
key                 value                                                                             
=================== ==================================================================================
AcquireTime         [{'value': 0.001}, {'value': pvaccess.pvaccess.ScalarType.DOUBLE}]                
AttributesFileParam [{'value': '/tmp/attributes.xml'}, {'value': pvaccess.pvaccess.ScalarType.STRING}]
CameraManufacturer  [{'value': 'Simulated detector'}, {'value': pvaccess.pvaccess.ScalarType.STRING}] 
CameraModel         [{'value': 'Basic simulator'}, {'value': pvaccess.pvaccess.ScalarType.STRING}]    
ColorMode           [{'value': 0}, {'value': pvaccess.pvaccess.ScalarType.INT}]                       
HDF5FrameLocation   [{'value': '/entry/data/data'}, {'value': pvaccess.pvaccess.ScalarType.STRING}]   
ImageCounter        [{'value': 731420}, {'value': pvaccess.pvaccess.ScalarType.INT}]                  
codec               ========== ===========================================================            
                    key        value                                                                  
                    ========== ===========================================================            
                    name                                                                              
                    parameters ({'value': 5}, {'value': pvaccess.pvaccess.ScalarType.INT})            
                    ========== ===========================================================            
uncompressedSize    1048576                                                                           
uniqueId            731420                                                                            
=================== ==================================================================================

(base) prjemian@zap:~/.../BCDA-APS/bluesky_training$ 
```

**Review by stevehenke — APPROVED** [review]
This example demonstrates a useful capability and is easy to follow. Nice job, Pete!

File: examples/pva_examiner.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    print(pva_to_table(pv_object))
+    metadata = get_pva_ndattributes(pv_object)
+    print("NDAttributes (metadata)")
+    print(dictionary_to_table(metadata))
+
+
+def endless():
+    """Monitor the PVA for updates.  Print tables."""
+    channel = pvaccess.Channel(LISTEN_PVA)
+    channel.subscribe("monitor", monitor)
```
Comment (sveseli) [inline examples/pva_examiner.py:88]: When there is a single subscriber, monitor() call can be used to subscribe and start monitoring channel.

**Review by tguruswamy — APPROVED** [review]
Useful example, I ran it without issue. Do we have any examples of PVA channels in existing IOCs besides Pva1:Image ?

**Comment by prjemian** [comment]
> Useful example, I ran it without issue. Do we have any examples of PVA channels in existing IOCs besides Pva1:Image ?

At beam lines, PVA is still very new.  @sveseli Any examples from accelerator controls?

**Comment by sveseli** [comment]
DAQ software uses PVA in all IOCs, not sure what are we looking for exactly...

**Comment by prjemian** [comment]
> DAQ software uses PVA in all IOCs, not sure what are we looking for exactly...

Any PVA PV that could be monitored by this example.

**Comment by prjemian** [comment]
Just _how general_ is this example?

**Comment by sveseli** [comment]
So this works with NTNDArrays obviously, but the same approach works with any object. If you want to inspect structure of a particular channel, you can invoke getStructureDict() method on the received PvObject. The daq-aggregator (see https://git.aps.anl.gov/C2/daq/apps/daq-aggregator) for example, can monitor any number of CA and/or PVA channels and create/serve aggregate PV Object.

**Comment by prjemian** [comment]
> Just _how general_ is **this example**?

**Comment by sveseli** [comment]
Sorry, I am not really sure what you are looking for... The code you got should work with NTNDArrays. If you want to test whether it works in general, you can point it to any CA channel and see what happens, just make sure you add correct provider type.

**Comment by prjemian** [comment]
Thanks.  I'll need to add some advice on how to set the provider type, then.

**Comment by sveseli** [comment]
https://epics.anl.gov/extensions/pvaPy/production/pvaccess.html#channel

**Comment by sveseli** [comment]
Also, if you want a simple test server for something other than AD images, you can use something like this: 

https://git.aps.anl.gov/C2/daq/apps/daq-aggregator/-/blob/master/daq_aggregator/pvaTestServer.py

If you want, you can also easily create CA IOC in python: https://github.com/epics-base/pvaPy/blob/master/examples/caIocExample.py

**Review by keenanlang — APPROVED** [review]

**Comment by prjemian** [comment]
@sveseli @stevehenke @tguruswamy @keenanlang - Thanks for the reviews!  Your contributions have been added to the example code, either as new capability (such as PVA -or- CA via command line options) or comments.

**Pull request merged**